lib/charms/opensearch/v0/helper_secrets.py

```diff
@@ -0,0 +1,227 @@
+
+    @override
+    def get(
+        self,
+        scope: Scope,
+        key: str,
+        default: Optional[Union[int, float, str, bool]] = None,
+        auto_casting: bool = True,
+    ) -> Optional[Union[int, float, str, bool]]:
+        """Getting a secret's value."""
+        juju_version = JujuVersion.from_environ()
+
+        value = None
+        if juju_version.has_secrets:
+            if self.secrets(scope):
+                secret = self._juju_secret(scope, key)
+                if secret:
+                    start = time.time()
+                    value = secret.get_content().get(key)
+                    end = time.time()
+                    logging.info(f"Getting a secret's contents takes {end}-{start}")
+        else:
+            value = super().get(scope, key)
+
+        if not auto_casting:
+            return value
+
+        if value:
+            return self.cast(value)
+
+    @override
+    def get_object(self, scope: Scope, key: str) -> Optional[Dict[str, any]]:
+        """Get dict object from the relation data store."""
+        juju_version = JujuVersion.from_environ()
+
+        if not juju_version.has_secrets:
+            return super().get_object(scope, key)
+
+        result = {}
+        category = key
+        relation_data = self._get_relation_data(scope)
+        for secret_key in relation_data:
+            if f"{SECRET_LABEL_SEPARATOR}{category}{SECRET_LABEL_SEPARATOR}" not in secret_key:
+                continue
+
+            secret_id = relation_data.get(secret_key)
+            if not secret_id:
+                raise OpenSearchSecretNotFoundError(f"{category}:{secret_key}")
+
+            try:
+                start = time.time()
+                secret = self._charm.model.get_secret(id=secret_id)
+                end = time.time()
+                logging.info(f"Getting a secret takes {end}-{start}")
+            except SecretNotFoundError:
+                raise OpenSearchSecretNotFoundError(f"{category}:{secret_key}")
+
+            raw_content = secret.get_content()
+            if not raw_content:
+                raise OpenSearchSecretNotFoundError(f"{category}:{secret_key}")
+
+            ((raw_key, raw_value),) = raw_content.items()
+            real_key = raw_key.split(SECRET_LABEL_SEPARATOR)[-1]
+            result.update({real_key: raw_value})
+
+        if result:
+            return result
+
+    @override
+    def put(self, scope: Scope, key: str, value: Optional[Union[any]]) -> None:
+        """Adding or updating a secret's value."""
+        juju_version = JujuVersion.from_environ()
+
+        if not juju_version.has_secrets:
+            return super().put(scope, key, value)
+
+        relation_data = self._get_relation_data(scope)
+        start = time.time()
+        secret = self._scope_obj(scope).add_secret({key: value})
+        end = time.time()
+        logging.info(f"Adding secret takes {end}-{start}")
+        relation_data[self.label(scope, key)] = secret.id
```

Comment:
This code is added to the PR in the `[WIP]` stage so we had a clear confirmation on public pipelines.
(Similar to how it was done [here](https://warthogs.atlassian.net/browse/DPE-2258))

Reply:
Ah, sry, I thought you meant it for the time measurement above :sweat_smile: 

Generally it may be smart to save the secret ID somewhere... May relate for Juju to keep track of 'observers'.

NOTE to summarize offline discussions relating to the topic: latency or not -- we decided to cache data for the scope of events.